planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/MbtilesWriter.java

```diff
@@ -270,23 +271,30 @@ private void tileEncoder(Iterable<TileBatch> prev, Consumer<TileBatch> next) thr
         byte[] bytes, encoded;
         Integer tileDataHash;
         if (tileFeatures.hasSameContents(last)) {
+          if (skipFilled && lastIsFill) {
+            continue;
+          }
           bytes = lastBytes;
           encoded = lastEncoded;
           tileDataHash = lastTileDataHash;
           memoizedTiles.inc();
         } else {
           VectorTile en = tileFeatures.getVectorTileEncoder();
-          encoded = en.encode();
-          bytes = gzip(encoded);
+          if (skipFilled) {
+            lastIsFill = en.containsOnlyPolygonFills();
+            if (lastIsFill) {
+              continue;
+            }
+          }
+          lastEncoded = encoded = en.encode();
+          lastBytes = bytes = gzip(encoded);
           last = tileFeatures;
-          lastEncoded = encoded;
-          lastBytes = bytes;
           if (encoded.length > 1_000_000) {
             LOGGER.warn("{} {}kb uncompressed",
               tileFeatures.tileCoord(),
               encoded.length / 1024);
           }
-          if (compactDb && tileFeatures.getNumFeaturesToEmit() < MAX_FEATURES_HASHING_THRESHOLD) {
+          if (compactDb && en.containsOnlyPolygonFills()) {
```

Comment:
@msbarry hard to say w/o a benchmark for encoding but that seems a bit faster. I ran it for Australia twice on this branch and on the main branch. The downside is that the file size increases by almost 100MB since some opportunities to dedupe seem to be missed. So it depends on the goal whether to increase performance a bit or to reduce the file size.
```
# 1 skip-filled-tiles branch
    mbtiles          1m6s cpu:10m13s gc:4s avg:9.3
	  read    1x(22% 15s sys:2s wait:46s done:3s)
	  encode 11x(68% 45s wait:12s done:3s)
	  write   1x(57% 38s sys:4s wait:23s)
----------------------------------------
	features	3.6GB
	mbtiles	1.8GB

# 2 skip-filled-tiles branch
	mbtiles          1m7s cpu:10m23s gc:4s avg:9.3
	  read    1x(21% 14s sys:1s wait:46s done:3s)
	  encode 11x(68% 46s wait:12s done:3s)
	  write   1x(56% 38s sys:4s wait:24s)
----------------------------------------
	features	3.6GB
	mbtiles	1.8GB

# 3 main branch

	mbtiles          1m8s cpu:10m48s gc:4s avg:9.5
	  read    1x(10% 7s wait:58s done:3s)
	  encode 11x(72% 49s wait:10s done:3s)
	  write   1x(55% 37s sys:3s wait:26s)
----------------------------------------
	features	3.6GB
	mbtiles	1.7GB

# 4 main branch
	mbtiles          1m9s cpu:10m50s gc:4s avg:9.4
	  read    1x(10% 7s wait:58s done:3s)
	  encode 11x(72% 49s wait:11s done:3s)
	  write   1x(54% 37s sys:3s wait:26s)
----------------------------------------
	features	3.6GB
	mbtiles	1.7GB
```

Reply:
just saw your comment here https://github.com/onthegomap/planetiler/pull/234#issuecomment-1139539697 which suggests that it's even more compact than before https://github.com/onthegomap/planetiler/pull/219#issuecomment-1135880914
or was this before your latest commit https://github.com/onthegomap/planetiler/pull/234/commits/2d924c41e09eaf0a0e0727a767d141a121ce2138 which "fixed" some issue?

Reply:
oops, part of this might that I landed #235 but didn't merge it into this branch until just now - the water polygons make a big difference.

Reply:
okay, after rebasing the results are almost identical, so don't have any concerns then.
```
# 5 skip-filled-tiles branch rebased
    mbtiles          1m9s cpu:10m54s gc:4s avg:9.5
	  read    1x(9% 6s wait:58s done:3s)
	  encode 11x(72% 49s wait:10s done:3s)
	  write   1x(54% 37s sys:3s wait:26s)
----------------------------------------
	features	3.6GB
	mbtiles	1.7GB

# 6 skip-filled-tiles branch rebased
	mbtiles          1m9s cpu:10m50s gc:4s avg:9.4
	  read    1x(10% 7s wait:59s done:3s)
	  encode 11x(71% 49s wait:11s done:3s)
	  write   1x(53% 37s sys:3s wait:26s)
----------------------------------------
	features	3.6GB
	mbtiles	1.7GB
```

(it doesn't matter at all but just to mention: file size is 10MB larger with containsOnlyPolygonFills)

Reply:
@bbilger any preference here on `containsOnlyPolygonFills` vs. `containsOnlyPolygonFills && numFeatures < 5` (or a lower threshold?)

Reply:
```
cntFilled     =  636515
cntLt5        = 1422225
ctFilledLt5   =  635047
cntFilledGte5 =    1468
```
<details>
  <summary>Counted in #tileEncoder like so</summary>
  
```java
...
if (en.containsOnlyPolygonFills()) {
  cntFilled.incrementAndGet();
}
if (tileFeatures.getNumFeaturesToEmit() < 5) {
  cntLt5.incrementAndGet();
}
if (en.containsOnlyPolygonFills() && tileFeatures.getNumFeaturesToEmit() < 5) {
  ctFilledLt5.incrementAndGet();
}
if (en.containsOnlyPolygonFills() && tileFeatures.getNumFeaturesToEmit() >= 5) {
  cntFilledGte5.incrementAndGet();
}
if (compactDb && en.containsOnlyPolygonFills()) {
  tileDataHash = tileFeatures.generateContentHash();
} else {
  tileDataHash = null;
}
...
  ```
</details>

```sql
# containsOnlyPolygonFills
select count(*) from tile_data; -- 1290598

# tileFeatures.getNumFeaturesToEmit() < MAX_FEATURES_HASHING_THRESHOLD
select count(*) from tile_data; -- 1254643
```

To start with, I don't fully understand the motivation for replacing `tileFeatures.getNumFeaturesToEmit() < MAX_FEATURES_HASHING_THRESHOLD` by `containsOnlyPolygonFills`. The threshold check seems simple enough. But I assume you want to avoid unnecessary hash code generation & lookups.

So on the one hand: with the threshold approach we hash twice, and more than required but I think it should be fast enough, and we can remove a few more dupes (~10MB less).  
On the other hand with `containsOnlyPolygonFills` we miss a bunch of dupes, and this check also comes at a cost - but w/o a benchmark it's hard to say which is better/worse - would assume more or less identical overall.

So I'd tend to just keep `tileFeatures.getNumFeaturesToEmit() < MAX_FEATURES_HASHING_THRESHOLD`.  But if avoiding hash generation/lookup is the main motivation and the few missed dupes are neglectable, then I would go with `if (compactDb && tileFeatures.getNumFeaturesToEmit() < MAX_FEATURES_HASHING_THRESHOLD && en.containsOnlyPolygonFills())` (i.e. first check threshold to avoid uneccessary  #containsOnlyPolygonFills checks)

Reply:
The thought here was that "is fill" would be a more accurate indicator for whether a tile would be repeated a lot than the feature count heuristic. For example some future tileset where there are >5 polygons overlapping for a large portion of the planet.

For Australia (16,169,927 tiles), it looks like:

| # features | # tiles | repeated tiles | repeated (but not fill) tiles |
| -- | -- | -- | -- |
| 1 | 13,331,155 | 13,101,957 | 6,062 |
| 2 | 1,681,737 | 1,400,406 | 7,283 |
| 3 | 583,603 | 406,630 | 17,596 |
| 4 | 140,253 | 17,113 | 4,566 |
| 5 | 65,943 | 107 | 103 |
| 6 | 47,374 | 15 | 15 |
| 7 | 36,311 | 8 | 8 |
| 8 | 28,772 | 5 | 5 |
| 9 | 23,901 | 4 | 4 |
| >=10 | 230,878 | 31 | 31 |

So the isFill check catches 99.8% of repeated tiles and there end up being 41,350 distinct repeated tiles, but the <5 heuristic catches nearly 100% of repeated tiles, but there end up being 810,670 (20x more) distinct tiles. So I guess it's a tradeoff of whether we'd rather get the additional 0.2% to lower the mbtiles size slightly or if we want to minimize the chances of a hash collision by deduping 20x fewer tile hashes.

Reply:
Okay, thanks for the additional explanation and crunching more numbers.

So given that the "is fill"-check should be cheap&fast like you said, and
> For example some future tileset where there are >5 polygons overlapping for a large portion of the planet.

and 
> isFill check catches 99.8% of repeated tiles

and
> minimize the chances of a hash collision

are all very good argument to just go with your suggestion to use "is fill" instead - and keep the #features out of the equation entirely here.

Reply:
OK thanks, I was trying to decide whether `isFill() || numFeatures() <= N` would be better to catch more repeated tiles, but even just using the lowest threshold of 1 increases # distinct tiles being hashed to 267k (6-7x more) and only catches 17% of the remaining tiles, so it's probably not worth it.

If we wanted to catch more of the repeated tiles, it would probably be better to expand the isFill logic to also identify if the tile contains only edge of a rectangle or segment of a vertical/horizontal line (for example if a tileset includes latitude/longitude graticules).

> given that the "is fill"-check should be cheap&fast like you said

I double checked, visualvm shows 100ms on my macbook for australia and it's only taking ~100ms

Reply:
Looks like the "isEdge" check on lines and rectangle catches almost all of the other repeated tiles and only increases # distinct tiles to track to 53,491 (30% increase), so I'll add that in to be safe and cover future cases with graticules.